### PR TITLE
DCES-195 Get FDC files endpoint

### PR DESCRIPTION
**Pull request by LAA-Dave-Edwards**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/DCES-195)

Adding an endpoint that collects the data required for use for sending to the debt recovery team. 

## Checklist

Before you ask people to review this PR:

- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/mapper/FdcContributionsMapper.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import gov.uk.courtdata.dces.request.FdcContributionsRequest;
+import gov.uk.courtdata.entity.ContributionFilesEntity;
+import org.mapstruct.Mapper;
+import org.mapstruct.NullValuePropertyMappingStrategy;
+import org.mapstruct.ReportingPolicy;
+
+@Mapper(componentModel = "spring",
+        unmappedTargetPolicy = ReportingPolicy.IGNORE,
+        nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+public interface FdcContributionsMapper {
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/mapper/FdcContributionsMapper.java:12]: It looks like this class isn't used, can it be deleted?

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/request/FdcContributionsRequest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+import java.util.Set;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class FdcContributionsRequest {
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/request/FdcContributionsRequest.java:15]: It looks like this class is only used by `FdcContributionsMapper` , can this class be removed if not used?

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+@Service
+@RequiredArgsConstructor
+public class FdcContributionsService {
+    private final FdcContributionsRepository fdcContributionsRepository;
+
+    public List<FdcContributionsResponse> getFdcContributionFiles(FdcContributionsStatus status) {
+        log.info("Getting fdc contribution file with status with the -> {}", status);
+        final List<FdcContributionsEntity> fdcFileList = fdcContributionsRepository.findByStatus(status);
+
+        return fdcFileList.stream().map( cc -> FdcContributionsResponse.builder()
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/FdcContributionsService.java:24]: I would consider extracting the function as a named constant to aid readability, maybe something like this?

```
    private static final Function<FdcContributionsEntity, FdcContributionsResponse> FDC_ENTITY_TO_CONTRIBUTIONS_RESPONSE =
            cc -> FdcContributionsResponse.builder()
                    .id(cc.getId())
                    .finalCost(cc.getFinalCost())
                    .dateCalculated(cc.getDateCalculated())
                    .lgfsCost(cc.getLgfsCost())
                    .agfsCost(cc.getAgfsCost())
                    .sentenceOrderDate(Objects.nonNull(cc.getRepOrderEntity()) ? cc.getRepOrderEntity().getSentenceOrderDate() : null)
                    .build();
...
        return fdcFileList.stream()
                .map(FDC_ENTITY_TO_CONTRIBUTIONS_RESPONSE)
                .toList();
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+@XmlType
+public class FdcContributionsEntity {
+
+    @Id
+    @SequenceGenerator(name = "fdc_contributions_gen_seq", sequenceName = "S_GENERAL_SEQUENCE", allocationSize = 1, schema = "TOGDATA")
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "fdc_contributions_gen_seq")
+    @Column(name = "ID", nullable = false)
+    private Integer id;
+
+    @Column(name= "CONT_FILE_ID", nullable = false)
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java:31]: It looks like `CONT_FILE_ID` is nullable in the database, should it be marked `nullable = false` here? Or maybe the DB should have this constraint added?

File: maat-court-data-api/src/main/java/gov/uk/courtdata/enums/FdcContributionsStatus.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package gov.uk.courtdata.enums;
+
+public enum FdcContributionsStatus {
+    REQUESTED,
+    SENT,
+    REPLACED,
+    WAITING_ITEMS
+}
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/enums/FdcContributionsStatus.java:8]: Looking at the DB I think we are missing the status of `INVALID`

File: maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    @CreationTimestamp
+    @Column(name = "DATE_CREATED")
+    private LocalDate dateCreated;
+
+    @Builder.Default
+    @Column(name = "USER_CREATED")
+    private String userCreated = "DCES";
+
+    @Column(name = "USER_MODIFIED", length = 225)
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java:90]: I think the length should be only 100 here based on the DB schema.

> USER_MODIFIED	VARCHAR2(100 BYTE)	Yes		18

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/FdcContributionsRestControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$", hasSize(0)));
+    }
+
+    @Test
+    void testContributionFileContentWhenQueryParamIsNotProvided() throws Exception {
+        mvc.perform(MockMvcRequestBuilders.get(String.format(ENDPOINT_URL  +"/fdc-contribution-files"))
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isBadRequest());
```
Comment (stone-dj) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/FdcContributionsRestControllerTest.java:76]: Can you include an assertion on the body of the response also?

i.e.
`.andExpect(jsonPath("$.detail").value("Required parameter 'status' is not present."));`

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/FdcContributionsRestControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@WebMvcTest(FdcContributionsRestController.class)
+@AutoConfigureMockMvc(addFilters = false)
+class FdcContributionsRestControllerTest {
```
Comment (stone-dj) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/FdcContributionsRestControllerTest.java:25]: Can you include at a minimum a happy path and unhappy path integration test for the new controller also?

**Review by stone-dj — CHANGES_REQUESTED** [review]
I have left some feedback, please action it and let me know if I have misunderstood anything :)

File: maat-court-data-api/src/main/java/gov/uk/courtdata/enums/FdcContributionsStatus.java (hunk 5 above)
Comment (LAA-Dave-Edwards) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/enums/FdcContributionsStatus.java:8]: I had left out as it wasn't a use case we were looking into, but absolutely valid. Will add.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java (hunk 4 above)
Comment (LAA-Dave-Edwards) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/entity/FdcContributionsEntity.java:31]: It being nullable is the correct course of action.

**Review by stone-dj — APPROVED** [review]
Changes look good

**Review by vc13837 — APPROVED** [review]

**Pull request merged**